url43a0865bb0_kevinjdolan_darwin_tgz-pJ8-tui_DriverJ8/src/darwInvest/news/utility/Stemmer.java

```diff
@@ -0,0 +1,562 @@
+	private final boolean vowelinstem() {  
+		int i; 
+		for (i = 0; i <= j; i++) 
+			if (! cons(i)) 
+				return true;
+		return false;
+	}
+	
+	/* doublec(j) is true <=> j,(j-1) contain a double consonant. */
+	private final boolean doublec(int j) {  
+		if (j < 1) 
+			return false;
+		if (b[j] != b[j-1]) 
+			return false;
+		return cons(j);
+	}
+	
+	/* cvc(i) is true <=> i-2,i-1,i has the form consonant - vowel - consonant
+	      and also if the second c is not w,x or y. this is used when trying to
+	      restore an e at the end of a short word. e.g.
+	
+	         cav(e), lov(e), hop(e), crim(e), but
+	         snow, box, tray.
+	
+	 */
+	private final boolean cvc(int i) {  
+		if (i < 2 || !cons(i) || cons(i-1) || !cons(i-2)) 
+			return false;
+		int ch = b[i];
+		if (ch == 'w' || ch == 'x' || ch == 'y') 
+			return false;
+		return true;
+	}
+	
+	private final boolean ends(String s)
+	{  
+		int l = s.length();
+		int o = k-l+1;
+		if (o < 0) 
+			return false;
+		for (int i = 0; i < l; i++) 
+			if (b[o+i] != s.charAt(i)) 
+				return false;
+		j = k-l;
+		return true;
+	}
+	
+	/* setto(s) sets (j+1),...k to the characters in the string s, readjusting
+	      k. */
+	
+	private final void setto(String s) {  
+		int l = s.length();
+		int o = j+1;
+		for (int i = 0; i < l; i++) 
+			b[o+i] = s.charAt(i);
+		k = j+l;
+	}
+	
+	/* r(s) is used further down. */
+	
+	private final void r(String s) { 
+		if(m() > 0)
+			setto(s); 
+	}
+	
+	/* step1() gets rid of plurals and -ed or -ing. e.g.
+	
+	          caresses  ->  caress
+	          ponies    ->  poni
+	          ties      ->  ti
+	          caress    ->  caress
+	          cats      ->  cat
+	
+	          feed      ->  feed
+	          agreed    ->  agree
+	          disabled  ->  disable
+	
+	          matting   ->  mat
+	          mating    ->  mate
+	          meeting   ->  meet
+	          milling   ->  mill
+	          messing   ->  mess
+	
+	          meetings  ->  meet
+	
+	 */
+	private final void step1() {  
+		if (b[k] == 's') {  
+			if (ends("sses")) 
+				k -= 2; 
+			else if (ends("ies")) 
+				setto("i"); 
+			else if (b[k-1] != 's')
+				k--;
+		}
+		if (ends("eed")) { 
+			if (m() > 0) 
+				k--; 
+		} 
+		else if ((ends("ed") || ends("ing")) && vowelinstem()) {  
+			k = j;
+			if (ends("at")) 
+				setto("ate"); 
+			else if (ends("bl")) 
+				setto("ble"); 
+			else if (ends("iz")) 
+				setto("ize");
+			else if (doublec(k)) {  
+				k--;
+				int ch = b[k];
+				if (ch == 'l' || ch == 's' || ch == 'z')
+					k++;
+			}
+			else if (m() == 1 && cvc(k))
+				setto("e");
+		}
+	}
+	
+	/* step2() turns terminal y to i when there is another vowel in the stem. */
+	
+	private final void step2() {
+		if (ends("y") && vowelinstem()) 
+			b[k] = 'i'; 
+	}
+	
+	/* step3() maps double suffices to single ones. so -ization ( = -ize plus
+	      -ation) maps to -ize etc. note that the string before the suffix must give
+	      m() > 0. */
+	
+	private final void step3() { 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 24. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.